memtx: introduce the O(n) SK build (#11133)
tarantool/tarantool: changelogs/unreleased/gh-10847-memtx-sort-data.md

Reviewer: It'd be nice to have a perf test in `perf/lua` that could be used to estimate the benefit of the new approach.

Author: Done. See the last 2 commits.

Author: Haven't though about the tuple order consequences, fixed. Thanks.

<details>

<summary>diff</summary>

```
--- a/src/box/memtx_tree.cc
+++ b/src/box/memtx_tree.cc
@@ -2018,8 +2018,10 @@ memtx_tree_index_begin_build(struct index *base)
        struct memtx_engine *memtx = (struct memtx_engine *)base->engine;
        uint32_t space_id = base->def->space_id;
        uint32_t index_id = base->def->iid;
-       index->use_sk_sort_data = false;
-       if (memtx->sort_data_reader != NULL && index_id > 0 &&
+       index->use_sk_sort_data =
+               memtx->sort_data_reader != NULL && index_id > 0 &&
+               !space_has_before_replace_triggers(space_by_id(space_id));
+       if (index->use_sk_sort_data &&
            memtx_sort_data_reader_seek(memtx->sort_data_reader,
                                        space_id, index_id,
                                        &index->use_sk_sort_data) != 0) {
diff --git a/test/box-luatest/gh_10847_memtx_sort_data_test.lua b/test/box-luatest/gh_10847_memtx_sort_data_test.lua
index 87b804eee8..36d6e17033 100644
--- a/test/box-luatest/gh_10847_memtx_sort_data_test.lua
+++ b/test/box-luatest/gh_10847_memtx_sort_data_test.lua
@@ -127,6 +127,51 @@ g_generic.after_each(function(cg)
     cg.server:drop()
 end)
 
+-- Test that a before_replace trigger on the test space prevent using sort data.
+g_generic.test_before_replace = function(cg)
+    -- Create a snapshot and sort data file.
+    cg.server:exec(function()
+        box.snapshot()
+    end)
+
+    -- Load it with the _index.before_recovery_replace trigger set up.
+    local register_before_recovery_replace = [[
+        local trigger = require('trigger')
+        trigger.set('box.space.test.before_recovery_replace', 'test_trigger',
+                    function(old, new, space, req, header, body)
+                        return new
+                    end)
+    ]]
+    cg.server:restart({
+        env = {
+            ['TARANTOOL_RUN_BEFORE_BOX_CFG'] = register_before_recovery_replace,
+        }
+    })
+
+    -- Must not use the sort data because of the trigger.
+    local msg = 'Using MemTX sort data for building' ..
+                ' index \'sk\' of space \'test\''
+    t.assert(cg.server:grep_log(msg) == nil)
+
+    -- Now load it with a regular test.before_replace trigger set up.
+    local register_before_replace = [[
+        local trigger = require('trigger')
+        trigger.set('box.space.test.before_replace', 'test_trigger',
+                    function(old, new, space, req, header, body)
+                        return new
+                    end)
+    ]]
+    cg.server:restart({
+        env = {
+            ['TARANTOOL_RUN_BEFORE_BOX_CFG'] = register_before_replace,
+        }
+    })
+
+    -- Now it should use the sort data file (the regular before_replace trigger
+    -- does not fire on recovery so it should not affect the ability to use it).
+    t.assert(cg.server:grep_log(msg) ~= nil)
+end
+
 -- Test that a before_replace trigger on _index space prevent using sort data.
 g_generic.test_index_before_replace = function(cg)
     -- Create a snapshot and sort data file.
```

</details>

> ## feature/memtx

Author: Done.

> ## feature/memtx

Author: Done.

> * Now it's possible to sort secondary keys using a new O(n) sorting
>   algorithm using an additional data written into the snapshot. The
>   feature can be enabled with a new `memtx_sort_data_enabled` option
>   in `box.cfg` (gh-10847).

Reviewer 2: ```suggestion
* Now it is possible to sort secondary keys with a new O(n) sorting
  algorithm that uses additional data written into the snapshot. The
  feature can be enabled with a new `memtx_sort_data_enabled` option
  in `box.cfg` (gh-10847).
```
